### ISSUE TITLE
Unable to override

### ISSUE THREAD

**Issue by simonbyrne** [JuliaPackaging/BinaryBuilder.jl#625]
```
# ~/.julia/artifacts/Overrides.toml
[7cb0a576-ebde-5e09-9194-50597f1243b4]
MPICH = "/usr/local"
```

```julia
julia> using MPICH_jll, Libdl

julia> filter(x -> occursin("libmpi", x), dllist())
1-element Array{String,1}:
 "/usr/local/lib/libmpi.12.dylib"

julia> MPICH_jll.libmpi_path
"/usr/local/lib/libmpi.12.dylib"

julia> MPICH_jll.libmpi_handle
Ptr{Nothing} @0x00007fbbe3646cd0

julia> dlopen(libmpi)
ERROR: could not load library "@rpath/libmpi.12.dylib"
dlopen(@rpath/libmpi.12.dylib, 1): image not found
Stacktrace:
 [1] #dlopen#3(::Bool, ::typeof(dlopen), ::String, ::UInt32) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.3/Libdl/src/Libdl.jl:109
 [2] dlopen at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.3/Libdl/src/Libdl.jl:109 [inlined] (repeats 2 times)
 [3] top-level scope at REPL[3]:1

julia> libmpi
"@rpath/libmpi.12.dylib"
```

output of `otool -l /usr/local/lib/libmpi.12.dylib` gives
```
Load command 3
          cmd LC_ID_DYLIB
      cmdsize 72
         name /usr/local/opt/mpich/lib/libmpi.12.dylib (offset 24)
   time stamp 1 Wed Dec 31 16:00:01 1969
      current version 14.7.0
compatibility version 14.0.0
```

According to @staticfloat 
> No, it's that the LC_ID_DYLIB identifies itself as /usr/local/opt/mpich/lib/libmpi.12.dylib (edited) 
> So we would have to pass that string into ccall() and whatnot
> we can't pass @rpath/libmpi.12.dylib

There is also the issue that the libmpi may not be `libmpi.12.dylib` (e.g. if I wanted to use OpenMPI, then on my system it is libmpi.40.dylib.

cc: @staticfloat @giordano

**Comment by staticfloat** [comment]
We may be able to do something clever here, like have the JLL determine the ID, but I'm very wary of adding any kind of extra "smarts" to JLL packages; they're supposed to be as simple as possible, in order to make things easy to deal with.  I'm not 100% sold either way, but I'm thinking that we might have to push some complexity upstream and, instead of modifying the JLL package, force the MPI packages to be explicitly capable of dealing with arbitrary SONAMEs or mismatched dylib IDs.  I don't think making JLL packages happy to load different SONAMEs is really what we want to do.

From that perspective, using `Override.toml` might not even be the right thing to do; it might be better to have `OpenMPI.jl` or whatever the top-level driver is, deal with this explicitly, as this is a pretty strange situation.  I kind of don't want JLL packages to work with arbitrary libraries; they're supposed to provide a particular version of a particular library, and breaking that protocol, while it kinda-sorta works, seems like the wrong way forward, IMO.

I'm thinking we could have a switch somewhere that you can flip (compile-time environment variable, configuration TOML file somewhere, etc....) to tell `OpenMPI.jl` to try just `dlopen()`'ing a particular library name, and if it can be `dlopen()`'ed then it uses that system `libmpi`, rather than anything else.  Yes, it would require you to change from `ccall((libname, funcname), ...)` to `funcptr = dlsym(hdl, funcname); ccall(funcptr, ...)`.

**Comment by simonbyrne** [comment]
I'm sympathetic to that view: I was actually thinking of building a custom `MPI_jll` package that would fit in the middle, since (1) it would seem strange that you would override "MPICH" to point to a completely different MPI implementation, and  (2) [it seems unlikely that we will get MPICH to work on Windows](https://github.com/JuliaPackaging/Yggdrasil/pull/398) so we would probably have to use Microsoft MPI on that platform anyway. I haven't quite figured out how that would work yet.

@maleadt came up with a `dlsym` caching approach in the CUDA libraries that seems to reduce the overhead of dynamic `ccall`s:
https://github.com/JuliaGPU/CuArrays.jl/issues/488#issuecomment-552344283

Alternatively, it seems to me that the main reason we're jumping through all these hoops is that we don't have a good way to invalidate the precompile cache to a package if the library changes (since we only allow invalidation based on file timestamps). Is there a way we could add additional checks? We will probably need something like this for MPI, since there isn't a single ABI, and so need to define constants and type aliases based on the library.

**Comment by maleadt** [comment]
> Alternatively, it seems to me that the main reason we're jumping through all these hoops is that we don't have a good way to invalidate the precompile cache to a package if the library changes (since we only allow invalidation based on file timestamps). Is there a way we could add additional checks?

I [used to do that with CUDAdrv](https://github.com/JuliaGPU/CUDAdrv.jl/blob/eb5cf8e51f3e3cc828730ad91ea4894a00157c92/src/CUDAdrv.jl#L74-L78): invalidating the precompilation cache when the underlying CUDA library upgraded. But since that would mean precompiling on, e.g., compute nodes all the time, I decided to move the other direction (which includes the `@runtime_ccall` feature you linked to).

**Comment by simonbyrne** [comment]
I hadn't thought of deleting the cache file in the module init, I'm impressed that seems to work.

**Comment by vchuravy** [comment]
> I hadn't thought of deleting the cache file in the module init, I'm impressed that seems to work.

I would caveat against it. It was one of the bigger pain points of running the CUDA stack on HPC systems.

>  I was actually thinking of building a custom MPI_jll package that would fit in the middle

This seems kinda the right approach, make MPICH_jll and OpenMPI_jll lazy, allow for a broader Overrides.toml to work and support Windows MPI.

This is one of the reasons why I was hesitant about shipping MPI through BB. There is a strong forcing functions on HPC systems to use the vendor provided one and that needs to be as easy as possible.

**Comment by simonbyrne** [comment]
> This is one of the reasons why I was hesitant about shipping MPI through BB. There is a strong forcing functions on HPC systems to use the vendor provided one and that needs to be as easy as possible.

I agree entirely, but there is a slight caveat: often HPC systems have multiple MPI implementations (i.e. different versions or configuration options) and users will switch between them. Currently doing that requires a manual `Pkg.build`; only retriggering `precompile` is net improvement. I don't see how we can get rid of that altogether due to the multiple ABI issue.

**Comment by oschulz** [comment]
From more user-centric point of view: I believe a clear mechanism to opt for a specifc system-wide binary, ideally one the works for most packages, would be invaluable. On HPC systems, it's of course often necessary to load specific versions of CUDA, and especially MPI. But in other contexts this is important too: We often build containers that pack all deps that we may need, e.g. HDF5. Julia used to use that HDF5 (which is built with specific options), but now it always brings it's own.

There is another aspect: Disk space and file access. My "artifacts" directory is 2.2 GB now - in the past, I had almost no binaries in my ".julia", which kept it small and fast, because most packages used system-wide binaries when detected. With larger user groups, this blows up everyone's home directory significantly, and 2 GB is a size at which computing centers start to care, since larger data is usually not supposed to be kept in "$HOME". Also, "$HOME" is usually not suitable for fast parallel access. But putting artifacts on a cluster file system is also not great - there are about 50000 files in my "artifacts" dir, cluster file systems don't like that at all. On the other hand, when I can put all deps in a container, all compute nodes will essentially mount that as a remote block device, and all metadata operations become local. So a single container file can happily live on the cluster file system (or similar storage).

I realize that an installation experience that is as foolproof as possible is essential for the average user. But for advanced cases, an easy mechanism to select binaries for most (if not all) packages that have bin deps seems essential to me.

**Comment by simonbyrne** [comment]
> But for advanced cases, an easy mechanism to select binaries for most (if not all) packages that have bin deps seems essential to me.

I agree, and this is the idea behind Overrides.toml: the question is what interface the overrides need to provide.
